packages/models-tree/src/models-tree/ModelsTreeDefinition.ts

```diff
@@ -611,120 +617,140 @@ function createCategoryInstanceKeyPaths(categoryId: Id64String, idsCache: Models
-
     const ctes = [
-      `ModelsCategoriesElementsHierarchy(ECInstanceId, ParentId, ModelId, Path) AS (
+      `InstanceElementsWithClassGroupingNodes(ECInstanceId, ECClassId, ParentId, ModelId, CategoryId, GroupingNodeIndex) AS (
+        ${[...(targetElementsInfoQuery ? [targetElementsInfoQuery] : []), ...targetGroupingNodesElementInfoQueries].join(" UNION ALL ")}
+      )`,
+      `ModelsCategoriesElementsHierarchy(ECInstanceId, ParentId, ModelId, GroupingNodeIndex, Path) AS (
         SELECT
           e.ECInstanceId,
-          e.Parent.Id,
-          e.Model.Id,
-          json_array(
-            ${ECSql.createInstanceKeySelector({ alias: "e" })},
-            IIF(e.Parent.Id IS NULL, ${ECSql.createInstanceKeySelector({ alias: "c" })}, NULL),
-            IIF(e.Parent.Id IS NULL, ${ECSql.createInstanceKeySelector({ alias: "m" })}, NULL)
+          e.ParentId,
+          e.ModelId,
+          e.GroupingNodeIndex,
+          IIF(e.ParentId IS NULL,
+            ec_classname([m].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([m].[ECInstanceId]) AS TEXT) || '${separator}' || ec_classname([c].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([c].[ECInstanceId]) AS TEXT) || '${separator}' || ec_classname([e].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([e].[ECInstanceId]) AS TEXT),
+            ec_classname([e].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([e].[ECInstanceId]) AS TEXT)
           )
-        FROM ${hierarchyConfig.elementClassSpecification} e
-        JOIN bis.GeometricModel3d m ON m.ECInstanceId = e.Model.Id
-        JOIN bis.SpatialCategory c ON c.ECInstanceId = e.Category.Id
-        WHERE ${whereClause}
+
+        FROM InstanceElementsWithClassGroupingNodes e
+         LEFT JOIN bis.GeometricModel3d m ON (e.ParentId IS NULL AND m.ECInstanceId = e.ModelId)
+         LEFT JOIN bis.SpatialCategory c ON (e.ParentId IS NULL AND c.ECInstanceId = e.CategoryId)
 
         UNION ALL
 
         SELECT
           pe.ECInstanceId,
           pe.Parent.Id,
           pe.Model.Id,
-          json_insert(
-            ce.Path,
-            '$[#]', ${ECSql.createInstanceKeySelector({ alias: "pe" })},
-            '$[#]', IIF(pe.Parent.Id IS NULL, ${ECSql.createInstanceKeySelector({ alias: "c" })}, NULL),
-            '$[#]', IIF(pe.Parent.Id IS NULL, ${ECSql.createInstanceKeySelector({ alias: "m" })}, NULL)
+          ce.GroupingNodeIndex,
+          IIF(pe.Parent.Id IS NULL,
+            ec_classname([m].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([m].[ECInstanceId]) AS TEXT) || '${separator}' || ec_classname([c].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([c].[ECInstanceId]) AS TEXT) || '${separator}' || ec_classname([pe].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([pe].[ECInstanceId]) AS TEXT) || '${separator}' || ce.Path,
+            ec_classname([pe].[ECClassId], 's.c') || '${separator}' || CAST(IdToHex([pe].[ECInstanceId]) AS TEXT) || '${separator}' || ce.Path
```

Comment:
I'm not a big fan of such string concatenations... [Your comment](https://github.com/iTwin/viewer-components-react/pull/1084#issue-2641797375) says this way is slightly faster than JSON, but do we know why exactly? Maybe it's just because this is less verbose? In that case, we could make JSON less verbose too:

- Instead of using instance key objects in an array, we could use tuples - that would remove "className" and "id" strings from the output (`{"className":"x", "id":"y"}` becomes `["x", "y"]`).

- Maybe we don't need to exact full class name? I think it would be enough for us to know if it's model, category or element id. So instead of adding full class name, we could add "m", "c", "e" strings and, when parsing the path, convert them to "BisCore.Model", "BisCore.SpatialCategory", "BisCore.GeometricElement3d".

Reply:
Applied changes, they helped improve the performance. However, string concatenation still outperformed json_array version by about 1 second. Keeping string concatenation.